vendor/bat-native-ads/src/bat/ads/internal/ml/pipeline/text_processing/embedding_processing.cc

```diff
@@ -0,0 +1,131 @@
+#include "bat/ads/internal/base/strings/string_strip_util.h"
+#include "bat/ads/internal/ml/data/text_data.h"
+#include "bat/ads/internal/ml/data/vector_data.h"
+#include "bat/ads/internal/ml/pipeline/pipeline_embedding_info.h"
+#include "bat/ads/internal/ml/pipeline/pipeline_util.h"
+#include "bat/ads/internal/ml/pipeline/text_processing/embedding_data.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+
+namespace ads {
+namespace ml {
+namespace pipeline {
+
+// static
+std::unique_ptr<EmbeddingProcessing> EmbeddingProcessing::CreateFromValue(
+    base::Value resource_value,
+    std::string* error_message) {
+  DCHECK(error_message);
+
+  auto text_processing = std::make_unique<EmbeddingProcessing>();
+  if (!text_processing->FromValue(std::move(resource_value))) {
+    *error_message = "Failed to parse text embedding pipeline JSON";
+    return {};
+  }
+
+  return text_processing;
+}
+
+bool EmbeddingProcessing::IsInitialized() const {
+  return is_initialized_;
+}
+
+EmbeddingProcessing::EmbeddingProcessing() : is_initialized_(false) {}
+
+EmbeddingProcessing::~EmbeddingProcessing() = default;
+
+void EmbeddingProcessing::SetInfo(const PipelineEmbeddingInfo& info) {
+  embedding_pipeline_.version = info.version;
+  embedding_pipeline_.timestamp = info.timestamp;
+  embedding_pipeline_.locale = info.locale;
+  embedding_pipeline_.embeddings_dim = info.embeddings_dim;
+  embedding_pipeline_.embeddings = info.embeddings;
+}
+
+bool EmbeddingProcessing::FromValue(base::Value resource_value) {
+  absl::optional<PipelineEmbeddingInfo> pipeline_embedding_info =
+      ParsePipelineEmbedding(std::move(resource_value));
+
+  if (pipeline_embedding_info.has_value()) {
+    SetInfo(pipeline_embedding_info.value());
+    is_initialized_ = true;
+  } else {
+    is_initialized_ = false;
+  }
+
+  return is_initialized_;
+}
+
+std::string EmbeddingProcessing::CleanText(const std::string& text,
+                                           bool is_html) {
+  std::string cleaned_text = text;
+  if (is_html) {
+    cleaned_text = ParseTagAttribute(cleaned_text, "og:title", "content");
+  }
+  cleaned_text = StripNonAlphaCharacters(cleaned_text);
+  std::transform(cleaned_text.begin(), cleaned_text.end(), cleaned_text.begin(),
+                 ::tolower);
+  return cleaned_text;
+}
+
+TextEmbeddingData EmbeddingProcessing::EmbedText(
+    const std::string& text) const {
+  std::vector<float> embedding_initialize(embedding_pipeline_.embeddings_dim,
+                                          0.0);
+  VectorData embedding_vector = VectorData(embedding_initialize);
+  TextEmbeddingData embedding_data;
+  embedding_data.embedding = embedding_vector;
+
+  if (!IsInitialized()) {
+    return embedding_data;
+  }
+
+  const std::vector<std::string> tokens = base::SplitString(
+      text, " ", base::TRIM_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
+
+  std::string in_vocab_text;
+  float n_tokens = 0.0;
```

Comment:
int as we only increment by 1.0

Reply:
I kept this as a float due to the division operation used afterwards

Reply:
could we not static_cast<float> for the division. Just seems odd to have a count with float

Reply:
Changed to `int` and added a `static_cast<float>`